hydroflow_lang/src/graph/flat_graph.rs

```diff
@@ -17,46 +19,60 @@ use super::di_mul_graph::DiMulGraph;
 use super::partitioned_graph::PartitionedGraph;
 use super::{GraphEdgeId, GraphNodeId, Node, PortIndexValue};
 
-/// A graph representing a hydroflow dataflow graph before subgraph partitioning, stratification, and handoff insertion.
-/// I.e. the graph is a simple "flat" without any subgraph heirarchy.
-///
-/// May optionally contain handoffs, but in this stage these are transparent and treated like an identity operator.
-///
-/// Use `Self::into_partitioned_graph()` to convert into a subgraph-partitioned & stratified graph.
 #[derive(Debug, Default)]
-pub struct FlatGraph {
-    /// Each node (operator or handoff).
-    pub(crate) nodes: SlotMap<GraphNodeId, Node>,
-    /// Graph
-    pub(crate) graph: DiMulGraph<GraphNodeId, GraphEdgeId>,
-    /// Input and output port for each edge.
-    pub(crate) ports: SecondaryMap<GraphEdgeId, (PortIndexValue, PortIndexValue)>,
+pub struct FlatGraphBuilder {
     /// Spanned error/warning/etc diagnostics to emit.
-    pub(crate) diagnostics: Vec<Diagnostic>,
+    diagnostics: Vec<Diagnostic>,
 
-    /// What variable name each graph node belongs to (if any).
-    pub(crate) node_varnames: SparseSecondaryMap<GraphNodeId, Ident>,
+    /// FlatGraph being built.
+    flat_graph: FlatGraph,
     /// Variable names, used as [`HfStatement::Named`] are added.
     varname_ends: BTreeMap<Ident, Ends>,
+    /// Each (out -> inn) link inputted.
+    links: Vec<Ends>,
+
+    /// If the builder has been finalized/sealed.
+    finalized: bool,
 }
 
-impl FlatGraph {
-    /// Creates a new `FlatGraph` instance based on the [`HfCode`] AST.
-    ///
-    /// TODO(mingwei): better error/diagnostic handling. Maybe collect all diagnostics before emitting.
+impl FlatGraphBuilder {
+    /// Create a new empty graph builder.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
     pub fn from_hfcode(input: HfCode) -> Self {
-        let mut graph = Self::default();
+        input.into()
+    }
 
-        for stmt in input.statements {
-            graph.add_statement(stmt);
-        }
+    /// Finalize the builder, and emit any diagnostics.
+    /// Emits diagnostics, returns true if there are NO errors.
+    pub fn finalize(&mut self, min_diagnostic_level: Level) -> bool {
+        self.finalized = true;
 
-        graph.check_operator_errors();
-        graph
+        self.connect_operator_links();
```

Comment:
(optional) It'd be nice if we could have the stateful part of a `FlatGraphBuilder` only capture things the user has added, and have the finalization / resolution logic operate on local mutable state.

Alternatively, if we combine `finalize` / `build` that could take a `(self, ...)` instead of `(&mut self, ...)` since there are no guarantees on APIs working correctly after finalization?

Reply:
Did the later